File: reference-implementation/python/src/osp/paid_provisioning.py
"""Paid provisioning helpers for the Python OSP SDK.

Provides typed estimate models, payment proof envelope support, and
async polling helpers for paid provisioning flows.
"""

from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Awaitable, Callable, Optional

from pydantic import BaseModel, Field

from .types import Currency, PaymentMethod


# ---------------------------------------------------------------------------
# Payment Proof Envelope
# ---------------------------------------------------------------------------

class PaymentProofEnvelope(BaseModel):
    """Structured payment proof for Sardis or other payment rails."""

    version: str = "1"
    mandate_id: str
    amount: str
    currency: str
    provider_id: str
    offering_id: str
    tier_id: str
    signature: str
    expires_at: str
    escrow_id: Optional[str] = None
    nonce: Optional[str] = None

    def serialize(self) -> str:
        """Serialize to JSON string for ProvisionRequest.payment_proof."""
        return self.model_dump_json()

    @classmethod
    def parse(cls, raw: str) -> "PaymentProofEnvelope":
        """Parse a JSON string back into a typed envelope."""
        return cls.model_validate_json(raw)

    def is_expired(self) -> bool:
        """Check whether this proof has expired."""
        exp = datetime.fromisoformat(self.expires_at.replace("Z", "+00:00"))
        return exp < datetime.now(timezone.utc)


# ---------------------------------------------------------------------------
# Estimate Models
# ---------------------------------------------------------------------------

class EstimateCost(BaseModel):
    """Cost breakdown from an estimate response."""

    amount: str
    currency: Currency
    interval: Optional[str] = None


class EstimateResponse(BaseModel):
    """Typed response from POST /osp/v1/estimate."""

    offering_id: str
    tier_id: str
    cost: Optional[EstimateCost] = None
    accepted_payment_methods: list[str] = Field(default_factory=list)
    escrow_required: bool = False
    approval_required: bool = False
    estimated_provision_seconds: Optional[int] = None


class EstimateDecision(BaseModel):
    """Result of evaluating an estimate for payment decisions."""

    estimate: EstimateResponse
    requires_payment: bool
    requires_escrow: bool
    requires_approval: bool
    suggested_payment_method: Optional[PaymentMethod] = None


def evaluate_estimate(estimate: EstimateResponse) -> EstimateDecision:
    """Evaluate an estimate response and produce a payment decision.

    This is the recommended entry point before calling provision() on
    paid tiers.
    """
    is_free = (
        estimate.cost is None
        or estimate.cost.amount in ("0", "0.00", "0.000")
    )

    methods = estimate.accepted_payment_methods
    suggested: Optional[PaymentMethod] = None
    if is_free:
        suggested = PaymentMethod.free
    else:
        for m in methods:
            if m != "free":
                try:
                    suggested = PaymentMethod(m)
                except ValueError:
                    pass
                break

    return EstimateDecision(
        estimate=estimate,
        requires_payment=not is_free,
        requires_escrow=estimate.escrow_required,
        requires_approval=estimate.approval_required,
        suggested_payment_method=suggested,
    )


# ---------------------------------------------------------------------------
# Provision Request Builder
# ---------------------------------------------------------------------------

def build_paid_provision_request(
    decision: EstimateDecision,
    *,
    project_name: str,
    nonce: str,
    idempotency_key: str,
    proof: Optional[PaymentProofEnvelope] = None,
    region: Optional[str] = None,
    config: Optional[dict[str, Any]] = None,
) -> dict[str, Any]:
    """Build a provision request dict from an estimate decision.

    For free tiers, no proof is needed. For paid tiers, the proof
    envelope is serialized and attached.
    """
    request: dict[str, Any] = {
        "offering_id": decision.estimate.offering_id,
        "tier_id": decision.estimate.tier_id,
        "project_name": project_name,
        "nonce": nonce,
        "idempotency_key": idempotency_key,
    }

    if region:
        request["region"] = region
    if config:
        request["config"] = config

    if decision.requires_payment:
        if proof is None:
            raise ValueError(
                "Payment proof is required for paid tiers. "
                f"Suggested method: {decision.suggested_payment_method}"
            )
        if proof.is_expired():
            raise ValueError("Payment proof has expired — generate a fresh proof")
        request["payment_method"] = (
            decision.suggested_payment_method.value
            if decision.suggested_payment_method
            else "sardis_wallet"
        )
        request["payment_proof"] = proof.serialize()
    else:
        request["payment_method"] = "free"

    return request


# ---------------------------------------------------------------------------
# Async Polling Helper
# ---------------------------------------------------------------------------

class PaidProvisionError(Exception):
    """Raised when paid provisioning fails."""

    def __init__(self, message: str, response: Optional[dict] = None):
        super().__init__(message)
        self.response = response


class ApprovalRequiredError(Exception):
    """Raised when provisioning requires human approval."""

    def __init__(self, message: str, response: dict):
        super().__init__(message)
        self.response = response


async def poll_paid_provision(
    poll_fn: Callable[[], Awaitable[dict[str, Any]]],
    *,
    max_polls: int = 30,
    poll_interval_seconds: float = 2.0,
    on_poll: Optional[Callable[[int, str], None]] = None,
) -> dict[str, Any]:
    """Poll for async paid provisioning completion.

    When a provider returns 202 Accepted for a paid provision, the
    agent must poll the status endpoint until the resource becomes
    active or the operation fails.

    Args:
        poll_fn: Async function that fetches the current resource status.
        max_polls: Maximum number of poll attempts.
        poll_interval_seconds: Seconds between polls.
        on_poll: Optional callback invoked on each poll with attempt and status.

    Returns:
        The final provision response when status is "active" or terminal.

    Raises:
        PaidProvisionError: On failure or timeout.
        ApprovalRequiredError: When human approval is required.
    """
    for attempt in range(1, max_polls + 1):
        response = await poll_fn()
        status = response.get("status", "unknown")

        if on_poll:
            on_poll(attempt, status)

        if status == "active":
            return response

        if status in ("failed", "deprovisioned") or "error" in response:
            error_msg = response.get("error", {}).get("message", status)
            raise PaidProvisionError(
                f"Paid provisioning failed: {error_msg}", response
            )

        if status == "approval_required":
            raise ApprovalRequiredError(
                "Provision requires human approval before proceeding",
                response,
            )

        await asyncio.sleep(poll_interval_seconds)

    raise PaidProvisionError(
        f"Paid provisioning timed out after {max_polls} polls"
    )

File: reference-implementation/typescript/src/paid-provisioning.ts
/**
 * Paid provisioning helpers for the TypeScript OSP SDK.
 *
 * Provides estimate-first flows, payment proof attachment, and async
 * paid provisioning retry handling.
 */

import type {
  EstimateRequest,
  EstimateResponse,
  PaymentMethod,
  ProvisionRequest,
  ProvisionResponse,
} from "./types.js";

// ---------------------------------------------------------------------------
// Payment Proof
// ---------------------------------------------------------------------------

/** Structured payment proof envelope for Sardis or other rails. */
export interface PaymentProofEnvelope {
  version: string;
  mandate_id: string;
  amount: string;
  currency: string;
  provider_id: string;
  offering_id: string;
  tier_id: string;
  signature: string;
  expires_at: string;
  escrow_id?: string;
  nonce?: string;
}

/** Serialize a proof envelope to the string format expected by ProvisionRequest. */
export function serializePaymentProof(proof: PaymentProofEnvelope): string {
  return JSON.stringify(proof);
}

/** Parse a payment proof string back into a typed envelope. */
export function parsePaymentProof(raw: string): PaymentProofEnvelope {
  const parsed = JSON.parse(raw);
  if (!parsed.version || !parsed.mandate_id || !parsed.signature) {
    throw new Error("Invalid payment proof: missing required fields");
  }
  return parsed as PaymentProofEnvelope;
}

/** Check whether a proof envelope has expired. */
export function isProofExpired(proof: PaymentProofEnvelope): boolean {
  return new Date(proof.expires_at) < new Date();
}

// ---------------------------------------------------------------------------
// Estimate-First Flow
// ---------------------------------------------------------------------------

/** Result of an estimate-first check, enriched with payment decision metadata. */
export interface EstimateDecision {
  estimate: EstimateResponse;
  requiresPayment: boolean;
  requiresEscrow: boolean;
  requiresApproval: boolean;
  suggestedPaymentMethod: PaymentMethod | null;
}

/**
 * Evaluate an estimate response and produce a payment decision.
 *
 * This is the recommended entry point before calling provision() on paid tiers.
 */
export function evaluateEstimate(estimate: EstimateResponse): EstimateDecision {
  const cost = estimate.cost;
  const isFree =
    !cost ||
    cost.amount === "0" ||
    cost.amount === "0.00" ||
    cost.amount === "0.000";

  const methods = estimate.accepted_payment_methods ?? [];
  const escrowRequired = estimate.escrow_required === true;

  // Pick the first non-free method if payment is required
  const suggestedPaymentMethod: PaymentMethod | null = isFree
    ? "free"
    : (methods.find((m: string) => m !== "free") as PaymentMethod) ?? null;

  return {
    estimate,
    requiresPayment: !isFree,
    requiresEscrow: escrowRequired,
    requiresApproval: false, // Approval is determined at provision time
    suggestedPaymentMethod,
  };
}

/**
 * Build a provision request from an estimate decision and payment proof.
 *
 * For free tiers, no proof is needed. For paid tiers, the proof envelope
 * is serialized and attached.
 */
export function buildPaidProvisionRequest(
  decision: EstimateDecision,
  opts: {
    projectName: string;
    nonce: string;
    idempotencyKey: string;
    proof?: PaymentProofEnvelope;
    region?: string;
    config?: Record<string, unknown>;
  },
): ProvisionRequest {
  const request: ProvisionRequest = {
    offering_id: decision.estimate.offering_id,
    tier_id: decision.estimate.tier_id,
    project_name: opts.projectName,
    nonce: opts.nonce,
    idempotency_key: opts.idempotencyKey,
    region: opts.region,
    config: opts.config,
  };

  if (decision.requiresPayment) {
    if (!opts.proof) {
      throw new Error(
        "Payment proof is required for paid tiers. " +
          `Suggested method: ${decision.suggestedPaymentMethod}`,
      );
    }
    if (isProofExpired(opts.proof)) {
      throw new Error("Payment proof has expired — generate a fresh proof");
    }
    request.payment_method =
      decision.suggestedPaymentMethod ?? ("sardis_wallet" as PaymentMethod);
    request.payment_proof = serializePaymentProof(opts.proof);
  } else {
    request.payment_method = "free";
  }

  return request;
}

// ---------------------------------------------------------------------------
// Async Paid Provisioning Retry
// ---------------------------------------------------------------------------

/** Options for async paid provisioning polling. */
export interface AsyncPaidProvisionOptions {
  /** Maximum number of poll attempts (default: 30). */
  maxPolls?: number;
  /** Interval between polls in ms (default: 2000). */
  pollIntervalMs?: number;
  /** Callback invoked on each poll with the current status. */
  onPoll?: (attempt: number, status: string) => void;
}

/**
 * Poll for async paid provisioning completion.
 *
 * When a provider returns 202 Accepted for a paid provision, the agent
 * must poll the status endpoint until the resource becomes active or
 * the operation fails.
 *
 * @param pollFn - Function that fetches the current resource status
 * @param opts - Polling configuration
 * @returns The final provision response when status is "active" or terminal
 */
export async function pollPaidProvision(
  pollFn: () => Promise<ProvisionResponse>,
  opts: AsyncPaidProvisionOptions = {},
): Promise<ProvisionResponse> {
  const maxPolls = opts.maxPolls ?? 30;
  const interval = opts.pollIntervalMs ?? 2000;

  for (let attempt = 1; attempt <= maxPolls; attempt++) {
    const response = await pollFn();

    opts.onPoll?.(attempt, response.status);

    if (response.status === "active") {
      return response;
    }

    if (
      response.status === "failed" ||
      response.status === "deprovisioned" ||
      response.error
    ) {
      throw new PaidProvisionError(
        `Paid provisioning failed: ${response.error?.message ?? response.status}`,
        response,
      );
    }

    if (response.status === "approval_required") {
      throw new ApprovalRequiredError(
        "Provision requires human approval before proceeding",
        response,
      );
    }

    // Still provisioning — wait and retry
    await new Promise((resolve) => setTimeout(resolve, interval));
  }

  throw new PaidProvisionError(
    `Paid provisioning timed out after ${maxPolls} polls`,
    null,
  );
}

// ---------------------------------------------------------------------------
// Error Classes
// ---------------------------------------------------------------------------

export class PaidProvisionError extends Error {
  constructor(
    message: string,
    public readonly response: ProvisionResponse | null,
  ) {
    super(message);
    this.name = "PaidProvisionError";
  }
}

export class ApprovalRequiredError extends Error {
  constructor(
    message: string,
    public readonly response: ProvisionResponse,
  ) {
    super(message);
    this.name = "ApprovalRequiredError";
  }
}
